Mathlib/Probability/CondJensen.lean

```diff
@@ -0,0 +1,577 @@
+    let f := denseSeq ↑sᶜ
+    have φc : ∀ (i : ℕ), ∃ (φ : E →L[𝕜] 𝕜) (c : ℝ),
+      (∀ a ∈ ball ↑(f i) (infDist ↑(f i) s), re (φ a) < c) ∧ ∀ b ∈ s, c ≤ re (φ b) := by
+        intro i
+        have di : Disjoint (ball ↑(f i) (infDist ↑(f i) s)) s := disjoint_ball_infDist
+        apply geometric_hahn_banach_open (convex_ball ↑(f i)  (infDist ↑(f i) s)) isOpen_ball hs₁ di
+    choose L c hLc using φc
+    exists L
+    exists c
+    constructor
+    · ext x
+      simp only [ge_iff_le, sub_nonneg, mem_iInter, mem_setOf_eq]
+      constructor
+      · apply Function.mtr
+        simp only [not_forall, not_le]
+        intro hx
+        have pos : 0 < (infDist x s) / 2 := by
+          apply div_pos
+          exact (IsClosed.notMem_iff_infDist_pos hs₂ hs₃).mp hx
+          exact Nat.ofNat_pos
+        have hfi : ∃ (i : ℕ), dist ↑(f i) x < ((infDist x s) / 2) := by
+          simp only [← mem_compl_iff] at hx
+          have : ((ball ⟨x, hx⟩ ((infDist x s) / 2)) ∩ (range f)).Nonempty := by
+            apply Dense.inter_open_nonempty (denseRange_denseSeq ↑sᶜ)
+            exact isOpen_ball; exact nonempty_ball.mpr pos
+          simp only [nonempty_def] at this
+          rcases this with ⟨b, hb⟩
+          simp only [mem_inter_iff, mem_ball, mem_range] at hb
+          rcases hb.2 with ⟨i, hi⟩; use i
+          rw (config := {occs := .pos [1]}) [← hi] at hb; exact hb.1
+        rcases hfi with ⟨i, hi⟩
+        rw [dist_comm] at hi
+        have hfix : infDist ↑(f i) s ≥  ((infDist x s) / 2) := by
+          apply le_of_not_gt
+          intro hp
+          rcases (infDist_lt_iff hs₃).mp hp with ⟨y, hy1, hy2⟩
+          have hxy : dist x y < infDist x s :=
+            calc
+              dist x y ≤ dist x ↑(f i) + dist ↑(f i) y := dist_triangle x ↑(f i) y
+              _ < (infDist x s) / 2 + (infDist x s) / 2 := add_lt_add hi hy2
+              _ = infDist x s := by simp only [add_halves]
+          exact notMem_of_dist_lt_infDist hxy hy1
+        have hxfi : x ∈ (ball ↑(f i) (infDist ↑(f i) s)) := by
+          rw [mem_ball]
+          calc
+            dist x ↑(f i) < (infDist x s) / 2 := hi
+            _ ≤ infDist ↑(f i) s := hfix
+        exists i; apply (hLc i).1; exact hxfi
+      · intro hx i
+        exact ((hLc i).2 x hx)
+    · intro ha j
+      cases le_or_gt (c j) 0 with
+      | inl hl =>
+        use f j; apply ne_of_lt; apply lt_of_lt_of_le
+        · have : (⇑re ∘ ⇑(L j)) ↑(f j) < c j := by
+            apply (hLc j).1; apply mem_ball_self
+            have : ↑(f j) ∈ sᶜ := (f j).property
+            exact (IsClosed.notMem_iff_infDist_pos hs₂ hs₃).mp this
+          exact this
+        · exact hl
+      | inr hr =>
+        simp only [nonempty_def] at hs₃
+        rcases hs₃ with ⟨x, hxs⟩; use x
+        apply ne_of_gt; apply lt_of_lt_of_le
+        · exact hr
+        · have : c j ≤ (⇑re ∘ ⇑(L j)) x := by
+            apply (hLc j).2; exact hxs
+          exact this
+
+/-- Lemma 1.2.9 for product spaces. -/
```

Comment:
That docstring does not explain what the lemma does, and refers to a lemma from an unknown source.